### ISSUE TITLE
What are the proper GitHub Actions permissions for this action?

### ISSUE THREAD

**Issue by BretFisher** [EnricoMi/publish-unit-test-result-action#161]
There's a workflow permissions model for GITHUB_TOKEN, which defaults to all write in a repo, but ideally, this Action documents the minimum permissions it needs. I can make a readme PR if we can confirm what of the permissions this needs (starting from none):

https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idpermissions

I haven't tested yet, but this is what I'm guessing based on the source code.

mainly, `checks: write` and `pull-requests: write`.

```yaml
    permissions:
      actions: none
      checks:  write
      contents: none
      deployments: none
      issues: none
      packages: none
      pull-requests: write
      repository-projects: none
      security-events: none
      statuses: none
```

**Comment by BretFisher** [comment]
OK here is my final permissions list with comments that worked using the default features of this action. Note that if you set any permissions to read/write, then everything else becomes `none`. I didn't check if any optional action settings would cause this to need more permissions. @EnricoMi if you agree I can create a PR to add this to readme.

```yaml
permissions:
  # creates a new check with unit test results
  checks:  write
  # needs to git diff the repo
  contents: read
  # find the pr for this commit
  issues: read
  # write artifacts and comment on pr related to this
  pull-requests: write
```

**Comment by EnricoMi** [comment]
> if you agree I can create a PR to add this to readme.

Sure, if you find this useful for users, go ahead.

> `# write artifacts and comment on pr related to this`

this action does not write artifacts itself, so `pull-requests: write` is not needed to "write artifacts"

**Comment by slavafomin** [comment]
Thank you @BretFisher, this is indeed a very useful information. I found myself in situation where I want to use the minimal possible permissions (best security practices) and I don't know what they actually should be.

I can confirm that these work correctly:

```yaml
permissions:
  contents: read
  issues: read
  pull-requests: read
  checks: write
```

**Comment by EnricoMi** [comment]
> I had to enable `write` for `issues` and `pull-requests`.

@Lugana707 are you sure `issues: write` is required? I reckon `pull-request: write` is sufficient. I would even suspect `issues: read` is not needed as only pull requests are read / written via the API.

**Comment by EnricoMi** [comment]
For reference, I found these API calls in code and respective permissions at https://docs.github.com/en/rest/reference/permissions-required-for-github-apps

| code example                                  | API call                                             | GitHub token permission |
|:----------------------------------------------|:-----------------------------------------------------|:------------------------|
| repo = gh.get_repo                            | GET /repos/{owner}/{repo}                            | metadata read           |
| repo.compare                                  | GET /repos/{owner}/{repo}/compare/{base...:head}     | contents read           |
| commit = repo.get_commit                      | GET /repos/{owner}/{repo}/commits/{sha}              | contents read           |
| commit.get_check_runs                         | GET /repos/{owner}/{repo}/commits/{sha}/check-runs   | contents read           |
| pull = gh.search_issues                       | GET /search/issues                                   | metadata read           |
| pull.get_commits                              | GET /repos/{owner}/{repo}/pulls/{number}/commits     | pull requests read      |
| pull_request.create_issue_comment             | POST /repos/{owner}/{repo}/issues/{number}/comments  | pull requests write     |
| pull.get_issue_comment(comment_id) | GET /repos/{owner}/{repo}/issues/comments/{id} | pull requests read     |
| comment.edit(body) | PATCH /repos/{owner}/{repo}/issues/comments/{id} | pull requests write     |

Depending on the options of this action used, you may not require some `write` permissions. For instance if `comment_mode: off`, the `pull-requests: write` permission is not needed and `pull requests: read` should be sufficient.

There are also calls to the GitHub GraphQL API, but I could not find which permissions / scopes these need:
- query repository.pullRequest.comments
- mutation minimizeComment

**Comment by dato** [comment]
I didn't `pull-request: write` in order for the commit to be allowed, _but I also_ needed `issues: read` because, without it, it seemed unable to get to the writing stage. Some kind of query seemed to require read-only access to issues; _perhaps_ that particular query/endpoint, subsumes PRs under issues, I don't know.

tl;dr I never needed `issues: write` but I did need `issues: read`.

**Comment by camillesf** [comment]
I didn't `pull-request: write` in order for the commit to be allowed, _but I also_ needed `issues: read` because, without it, it seemed unable to get to the writing stage. Some kind of query seemed to require read-only access to issues; _perhaps_ that particular query/endpoint, subsumes PRs under issues, I don't know.

tl;dr I never needed `issues: write` but I did need `issues: read`.

**Comment by EnricoMi** [comment]
I have [tested permissions in any possible configuration](https://github.com/EnricoMi/publish-unit-test-result-action/runs/4477169655?check_suite_focus=true#step:1:14) and the actions works fine with:

    checks: write
    pull-requests: write

@camillesf, @Lugana707 @slavafomin @BretFisher Did you see the additional need for `contents: read` and `issues: read` / `issues:write` in public or private or enterprise GitHub? On a regular PR or from a fork?

Can you share your particular action configuration?

**Comment by EnricoMi** [comment]
I can confirm that
```
permissions:
  checks: write
  pull-requests: write
```
is sufficient for public repositories and
```
permissions:
  checks: write
  contents: read
  issues: read
  pull-requests: write
```
is sufficient for private repositories.

With `comment_mode: off`, the `pull-requests: write` permission is not needed.

This is added to the README.md in #206.

**Comment by EnricoMi** [comment]
It looks like Github has reworked their permissions, only

    permissions:
      checks: write
      pull-requests: write

is needed now. The action works fine even with

    permissions:
      checks: write
      pull-requests: write
      contents: none
      issues: none
      actions: none

in a private repository on a dependabot branch.

Can someone here please verify the action works for them with those permissions
- in a public repo
- in a private repo
- in a fork branch
- in a dependabot branch

**Comment by jragingfury** [comment]
The minimum permissions that I could get things to work with are:
```
permissions:
  checks: write
  contents: read
  pull-requests: write
```
The `contents: read` permission seems to be required in a private repo to get the checkout action to work.

**Comment by EnricoMi** [comment]
Alright, things seem unchanged then. Maybe my fork setup was not representative. Thanks a lot for verifying!

**Comment by jasonleenaylor** [comment]
@EnricoMi I found that in our PR https://github.com/sillsdev/libpalaso/pull/1395 where we added permissions we unexpectedly needed `issues: write` has something changed?

**Comment by jasonleenaylor** [comment]
Oops, our mistake. Thank you @EnricoMi , sorry to waste your time.
